cmd/zpool/zpool_iter.c

```diff
@@ -521,30 +521,114 @@ vdev_run_cmd(vdev_cmd_data_t *data, char *cmd)
 			free(env[i]);
 }
 
+/* run our command based on it's full path */
+static int
+vdev_run_cmd_fullpath(char *fullpath, vdev_cmd_data_t *data)
+{
+	if (fullpath == NULL || fullpath[0] == '\0')
+		return (1);
+
+	/* Does the script exist in path? */
+	if (access(fullpath, X_OK) == 0) {
+		vdev_run_cmd(data, fullpath);
+		return (0);
+	}
+
+	return (1);
+}
+
+/* Run our command script */
+static int
+vdev_run_cmd_path(char *path, char *cmd, vdev_cmd_data_t *data)
+{
+	char fullpath[MAXPATHLEN];
+
+	if (path == NULL || path[0] == '\0')
+		return (1);
+
+	if (cmd == NULL || cmd[0] == '\0')
+		return (1);
+
+	if (snprintf(fullpath, sizeof (fullpath), "%s/%s", path, cmd) == -1)
+		return (1);
+
+	return (vdev_run_cmd_fullpath(fullpath, data));
+}
+
+/*
+ * Generate the search path for zpool iostat/status -c scripts.
+ * The string returned must be freed.
+ */
+char *
+zpool_get_cmd_searchpath(void)
+{
+	const char *env;
+	char *sp = NULL;
+
+	env = getenv("ZPOOL_SCRIPTS_PATH");
+	if (env != NULL) {
+		sp = strdup(env);
+		return (sp);
+	}
+
+	env = getenv("HOME");
+	if (env != NULL) {
+		if (asprintf(&sp, "%s/.zpool.d:%s",
+		    env, ZPOOL_SCRIPTS_DIR) != -1) {
+			return (sp);
+		}
+	}
+
+	if (asprintf(&sp, "%s", ZPOOL_SCRIPTS_DIR) != -1)
+		return (sp);
+
+	return (NULL);
+}
+
 /* Thread function run for each vdev */
 static void
 vdev_run_cmd_thread(void *cb_cmd_data)
 {
 	vdev_cmd_data_t *data = cb_cmd_data;
 	const char *sep = ",";
 	char *cmd = NULL, *cmddup, *rest;
-	char fullpath[MAXPATHLEN];
+	char *sp, *spdup, *dir;
 
 	cmddup = strdup(data->cmd);
 	if (cmddup == NULL)
 		return;
 
+	sp = zpool_get_cmd_searchpath();
+	if (sp == NULL) {
+		free(cmddup);
+		return;
+	}
+
 	rest = cmddup;
 	while ((cmd = strtok_r(rest, sep, &rest))) {
-		if (snprintf(fullpath, sizeof (fullpath), "%s/%s",
-		    ZPOOL_SCRIPTS_DIR, cmd) == -1)
+		/* if cmd starts with '/', assume full path */
+		if (cmd[0] == '/') {
+			vdev_run_cmd_fullpath(cmd, data);
 			continue;
+		}
 
-		/* Does the script exist in our zpool scripts dir? */
-		if (access(fullpath, X_OK) == 0)
-			vdev_run_cmd(data, fullpath);
+		spdup = strdup(sp);
+		if (spdup == NULL)
+			break;
+
+		dir = strtok(spdup, ":");
```

Comment:
Yep, completely forgot that was a thread. Thanks for the fixed up version. I was thinking of only calling `zpool_get_cmd_search_path` once and duplicating the result, but I think the game from that is negligible.